mutex/src/workflows.ts

```diff
@@ -0,0 +1,66 @@
+import * as workflow from '@temporalio/workflow';
+
+interface LockRequest {
+  initiatorId: string;
+  timeoutMs: number;
+}
+
+interface LockResponse {
+  releaseSignalName: string;
+}
+
+export const currentWorkflowIdQuery = workflow.defineQuery<string | null>('current-workflow-id');
+export const hasLockQuery = workflow.defineQuery<boolean>('hasLock');
+export const lockRequestSignal = workflow.defineSignal<[LockRequest]>('lock-requested');
+export const lockAcquiredSignal = workflow.defineSignal<[LockResponse]>('lock-acquired');
+
+export async function lockWorkflow(requests = Array<LockRequest>()) {
+  let currentWorkflowId: string | null = null;
+  workflow.setHandler(lockRequestSignal, (req: LockRequest) => {
+    requests.push(req);
+  });
+  workflow.setHandler(currentWorkflowIdQuery, () => currentWorkflowId);
+  while (workflow.workflowInfo().historyLength < 2000) {
+    await workflow.condition(() => requests.length > 0);
+    const req = requests.shift();
+    if (req === undefined) {
+      continue;
+    }
+    currentWorkflowId = req.initiatorId;
+    const handle = workflow.getExternalWorkflowHandle(req.initiatorId);
+    const releaseSignalName = workflow.uuid4();
+    await handle.signal(lockAcquiredSignal, { releaseSignalName });
+    let released = false;
+    workflow.setHandler(workflow.defineSignal(releaseSignalName), () => {
+      released = true;
+    });
+    await workflow.condition(() => released, req.timeoutMs);
+    currentWorkflowId = null;
+  }
+  // carry over any pending requests to the next execution
+  await workflow.continueAsNew<typeof lockWorkflow>(requests);
+}
+
+export async function testLockWorkflow(lockWorkflowId: string, sleepForMs = 500, lockTimeoutMs = 1000) {
```

Comment:
maybe name it `taskProcessor` or something indicating what you'd use a lock for

Reply:
I think `taskProcessor` is a little too generic, doesn't provide enough info about what this Workflow does. I'm fine with changing this, but I haven't been able to come up with a name that I like better than `testLockWorkflow`.

Reply:
Maybe `oneAtATimeWorkflow` that calls an activity `useAPIThatCantBeCalledInParallel()` 😄